sources/folderUtils.ts

```diff
@@ -1,10 +1,41 @@
-import {mkdirSync}        from 'fs';
-import {homedir, tmpdir}  from 'os';
-import {join}             from 'path';
-import type { NodeError } from './nodeUtils';
+import {existsSync, mkdirSync, renameSync} from 'fs';
+import {homedir, tmpdir}                   from 'os';
+import {join}                              from 'path';
+import process                             from 'process';
+
+import type {NodeError}                    from './nodeUtils';
 
 export function getInstallFolder() {
-  return process.env.COREPACK_HOME ?? join(homedir(), `.node/corepack`);
+  if (process.env.COREPACK_HOME == null) {
+    // TODO: remove this block on the next major.
+    const oldCorepackDefaultHome = join(homedir(), `.node`, `corepack`);
+    const newCorepackDefaultHome = join(
+      process.env.XDG_CACHE_HOME ??
+        process.env.LOCALAPPDATA ??
+        join(
+          homedir(),
+          process.platform === `win32` ? `AppData/Local` : `.cache`,
+        ),
+      `node/corepack`,
+    );
+    if (
+      existsSync(oldCorepackDefaultHome) &&
+      !existsSync(newCorepackDefaultHome)
+    ) {
+      mkdirSync(newCorepackDefaultHome, {recursive: true});
+      renameSync(oldCorepackDefaultHome, newCorepackDefaultHome);
+    }
+    return newCorepackDefaultHome;
+  }
+  return (
+    process.env.COREPACK_HOME ??
+    join(
+      process.env.XDG_CACHE_HOME ??
+        process.env.LOCALAPPDATA ??
+        join(homedir(), process.platform === `win32` ? `AppData/Local` : `.cache`),
+      `node/corepack`,
+    )
+  );
```

Comment:
You should swap the logic so that this part is the early return, since it's the shortest - also I think the `??` here is useless, because you already checked for `process.env.COREPACK_HOME == null` earlier 🤔

Reply:
The idea is to remove the block above in the next major (see TODO line 10). It's written like this so it's easier to remove the "migration path" that we definitely don't want to keep around for too long.

Reply:
Fair enough 👍